File: output/GPT4o-Mini/xCodeEval_4000/C++/Java/af3601be887567a34e783fe89388b257.java
import java.util.*;

public class af3601be887567a34e783fe89388b257_nc {
    static final int N = 3000 + 7;
    static int n, k;
    static int[] a = new int[N], b = new int[N];
    static List<Integer>[] g = new ArrayList[N];
    static Pair<Integer, Long>[][] dp = new Pair[N][N];
    static long[] val = new long[N];
    static int[] siz = new int[N];

    static {
        for (int i = 0; i < N; i++) {
            g[i] = new ArrayList<>();
            for (int j = 0; j < N; j++) {
                dp[i][j] = new Pair<>(Integer.MIN_VALUE, Long.MIN_VALUE);
            }
        }
    }

    static boolean read(Scanner scanner) {
        if (!scanner.hasNextInt()) return false;
        n = scanner.nextInt();
        k = scanner.nextInt();
        for (int i = 0; i < n; i++) g[i].clear();
        for (int i = 0; i < n; i++) a[i] = scanner.nextInt();
        for (int i = 0; i < n; i++) b[i] = scanner.nextInt();
        for (int i = 0; i < n - 1; i++) {
            int v = scanner.nextInt() - 1;
            int u = scanner.nextInt() - 1;
            g[v].add(u);
            g[u].add(v);
        }
        return true;
    }

    static void dfs(int v, int p) {
        val[v] = b[v] - a[v];
        dp[v][0] = new Pair<>(0, val[v]);
        siz[v] = 1;

        for (int u : g[v]) {
            if (u != p) {
                dfs(u, v);
                Pair<Integer, Long>[] tmp = new Pair[siz[v] + siz[u]];
                Arrays.fill(tmp, new Pair<>(Integer.MIN_VALUE, Long.MIN_VALUE));
                for (int i = 0; i < siz[v]; i++) {
                    for (int j = 0; j <= siz[u]; j++) {
                        Pair<Integer, Long> nw = new Pair<>(dp[v][i].first + dp[u][j].first, dp[v][i].second + dp[u][j].second);
                        tmp[i + j] = max(tmp[i + j], nw);
                    }
                }
                for (int i = 0; i < tmp.length; i++) {
                    dp[v][i] = tmp[i];
                }
                siz[v] += siz[u];
                val[v] += val[u];
            }
        }

        for (int i = siz[v] - 1; i >= 0; i--) {
            dp[v][i + 1] = max(dp[v][i + 1], new Pair<>(dp[v][i].first + (dp[v][i].second > 0 ? 1 : 0), 0L));
        }
    }

    static void solve() {
        dfs(0, -1);
        System.out.println(dp[0][k - 1].first + (dp[0][k - 1].second > 0 ? 1 : 0));
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int tc = scanner.nextInt();
        while (tc-- > 0) {
            if (!read(scanner)) break;
            solve();
        }
        scanner.close();
    }

    static Pair<Integer, Long> max(Pair<Integer, Long> a, Pair<Integer, Long> b) {
        if (a.first > b.first) return a;
        if (a.first < b.first) return b;
        return a.second > b.second ? a : b;
    }

    static class Pair<F, S> {
        F first;
        S second;

        Pair(F first, S second) {
            this.first = first;
            this.second = second;
        }
    }
}

File: output/GPT4o-Mini/xCodeEval_4000/C++/C/af3601be887567a34e783fe89388b257.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <math.h>

#define MAX_N 3007
#define INF INT_MAX
#define INF64 LLONG_MAX

typedef long long li;
typedef struct {
    int x;
    li y;
} Pair;

int n, k;
int a[MAX_N], b[MAX_N];
int g[MAX_N][MAX_N], g_size[MAX_N];

Pair dp[MAX_N][MAX_N];
li val[MAX_N];
int siz[MAX_N];

void dfs(int v, int p) {
    val[v] = b[v] - a[v];
    for (int i = 0; i <= n; i++) {
        dp[v][i].x = -INF;
        dp[v][i].y = -INF64;
    }
    dp[v][0].x = 0;
    dp[v][0].y = val[v];
    siz[v] = 1;

    for (int i = 0; i < g_size[v]; i++) {
        int u = g[v][i];
        if (u != p) {
            dfs(u, v);
            Pair tmp[MAX_N] = {{-INF, -INF64}};
            for (int j = 0; j < siz[v]; j++) {
                for (int l = 0; l <= siz[u]; l++) {
                    Pair nw = {dp[v][j].x + dp[u][l].x, dp[v][j].y + dp[u][l].y};
                    if (nw.x > tmp[j + l].x || (nw.x == tmp[j + l].x && nw.y > tmp[j + l].y)) {
                        tmp[j + l] = nw;
                    }
                }
            }
            for (int j = 0; j < siz[v] + siz[u]; j++) {
                dp[v][j] = tmp[j];
            }
            siz[v] += siz[u];
            val[v] += val[u];
        }
    }

    for (int i = siz[v] - 1; i >= 0; i--) {
        dp[v][i + 1].x = fmax(dp[v][i + 1].x, dp[v][i].x + (dp[v][i].y > 0));
        dp[v][i + 1].y = 0;
    }
}

void solve() {
    dfs(0, -1);
    printf("%d\n", dp[0][k - 1].x + (dp[0][k - 1].y > 0));
}

int read() {
    if (scanf("%d%d", &n, &k) != 2) return 0;
    for (int i = 0; i < n; i++) {
        g_size[i] = 0;
    }
    for (int i = 0; i < n; i++) {
        scanf("%d", &a[i]);
    }
    for (int i = 0; i < n; i++) {
        scanf("%d", &b[i]);
    }
    for (int i = 0; i < n - 1; i++) {
        int v, u;
        scanf("%d%d", &v, &u);
        v--; u--;
        g[v][g_size[v]++] = u;
        g[u][g_size[u]++] = v;
    }
    return 1;
}

int main() {
    int tc;
    scanf("%d", &tc);
    while (tc--) {
        if (!read()) break;
        solve();
    }
    return 0;
}
